compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+      if (node->isOtherwise()) {
+        ss_ << "otherwise ";
+      } else {
+        ss_ << "when ";
+        interpose(ss_, node->caseExprs(), ", ");
+        ss_ << " ";
+      }
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const While* node) {
+      ss_ << "while ";
+      printAst(ss_, node->condition());
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const WithClause* node) {
+      ss_ << "with ";
+      interpose(ss_, node->exprs(), ", ", "(", ")");
+    }
+
+    void visit(const Yield* node) {
+      ss_ << "yield ";
+      printAst(ss_, node->value());
+    }
+
+    void visit(const Zip* node) {
+      ss_ << "zip";
+      interpose(ss_, node->actuals(),", ","(",")");
+    }
+
+  };
+
+  /***************************************************************
+   * Specializations for generating userString when converting
+   * from uAST to old AST
+   */
+
+  struct UserStringVisitor {
+    std::ostringstream ss_;
+    UserStringVisitor(std::ostringstream ss_ = std::ostringstream()) {}
+
+
+    void visit(const Function* node) {
+      std::stringstream ss;
+      if (node->visibility() != Function::Visibility::DEFAULT_VISIBILITY) {
+        ss_ << kindToString(node->visibility()) << " ";
+      }
+
+      // storage kind
+      if (node->thisFormal() != nullptr && node->thisFormal()->storageKind() != IntentList::DEFAULT_INTENT) {
+        ss_ << kindToString(node->thisFormal()->storageKind()) <<" ";
+      }
+      // secondary methods
+      if (node->isMethod() && !node->isPrimaryMethod() && node->thisFormal()->child(0)->isIdentifier()) {
+        ss_ << node->thisFormal()->child(0)->toIdentifier()->name().str() << ".";
+      }
+
+      // Function Name
+      ss_ << node->name().str();
+
+      // Formals
+      int numThisFormal = node->thisFormal() ? 1 : 0;
+      int nFormals = node->numFormals() - numThisFormal;
+      if (nFormals == 0 && node->isParenless()) {
+        // pass
+      } else if (nFormals == 0) {
+        ss_ << "()";
+      } else {
+        auto it = node->formals();
+        interpose(ss_, it.begin() + numThisFormal, it.end(), ", ", "(", ")");
+      }
+
+      // Return type
+      if (const Expression* e = node->returnType()) {
+        ss_ << ": ";
+        printAst(ss_, e);
+      }
+
+      // Return Intent
+      if (node->returnIntent() != Function::ReturnIntent::DEFAULT_RETURN_INTENT) {
+        ss_ << " " << kindToString(node->returnIntent());
+      }
+
+    }
+
+    void visit(const ASTNode* node) {
+      ss_ << tagToString(node->tag());
+    }
+
+  };
+
+  /* Generic printer calling the above functions */
+  void printAst(std::ostream& os, const ASTNode* node) {
+    auto visitor = ChplSyntaxVisitor{};
+    node->dispatch<void>(visitor);
+    // TODO: Avoid making copy of string here
+    // challenge: when using rdbuf(), if nothing gets added to ss_, then
```

Comment:
Huh, is there no way to check the `ostream` and see if a write occurred? Perhaps get the original buffer state and record it?

Reply:
Seems like it wasn't working because `visitor.ss_` was an `ostringstream`. Changing it to `stringstream` allowed me to do the check as intended